src/nlv_object.h

```diff
@@ -48,36 +26,58 @@ class NLVObjectBasePtr
   friend class NLVObject;
 
 
-class NLVObjectBase : public Nan::ObjectWrap
+class NLVObjectBase : public Nan::ObjectWrap, public std::enable_shared_from_this<NLVObjectBase>
 {
 public:
+  void AddToParent(NLVObjectBase* parent) {
+    parent->PushChild(shared_from_this());
+  }
+  
+  void ClearChildren() {
+    for (auto& ptr: children_) {
+      if (auto object = ptr.lock()) {
+        object->ClearHandle();
+        object->ClearChildren();
+      }
+    }
+    children_.clear();
+  }
+
   virtual void ClearHandle() = 0;
-  virtual void ClearChildren() = 0;
-  virtual void SetParentReference(NLVObjectBasePtr *parentReference) = 0;
 
-  std::vector<NLVObjectBasePtr*> children_;
+  void PushChild(const std::shared_ptr<NLVObjectBase>& child) {
+    children_.emplace_front(child);
+  }
+  std::forward_list<std::weak_ptr<NLVObjectBase>> children_;
 };
 
 template <typename ParentClass, typename HandleType, typename CleanupHandler>
 class NLVObject : public NLVObjectBase
 {
+  std::shared_ptr<NLVObjectBase> selfPtr;
 public:
   typedef HandleType handle_type;
-
-  NLVObject(HandleType handle) : handle_(handle), parentReference_(NULL) {}
+  typedef typename std::remove_pointer<HandleType>::type HandleValue;
+    
+  NLVObject(HandleType handle) : handle_(handle, CleanupHandler::cleanup) {
+  }
+  
   ~NLVObject() {
-    // calling virtual ClearHandle() will break if overridden by subclasses
-    ClearHandle();
-    if (parentReference_ != NULL) {
-      parentReference_->SetInvalid();
-    }
+  }
+  
+  void RegisterSelf() {
+    selfPtr = shared_from_this();
   }
 
   static v8::Local<v8::Object> NewInstance(handle_type handle) {
     Nan::EscapableHandleScope scope;
     Local<Function> ctor = Nan::New<Function>(ParentClass::constructor);
     Local<Object> object = Nan::NewInstance(ctor).ToLocalChecked();
-    ParentClass *class_instance = new ParentClass(handle);
+    auto shared = std::shared_ptr<ParentClass>(new ParentClass(handle), [=](ParentClass*) {
+      // here we can now if GC has destroyed our object
+    });
+    ParentClass *class_instance = shared.get();
+    class_instance->RegisterSelf();
     class_instance->Wrap(object);
```

Comment:
I think in this case we shouldn't wrap this guy in a shared pointer. The idea of the smart pointer is to manage the lifetime of that object, assuming we have control over its lifetime. In this particular case we are about to give up ownership to v8 so it's a bit redundant here imho